src/server/svr_jobfunc.c

```diff
@@ -5408,6 +5410,15 @@ svr_histjob_update(job * pjob, int newstate, int newsubstate)
 	pjob->ji_qs.ji_substate = newsubstate;
 	set_statechar(pjob);
 
+	/* For subjob update the state */
+	if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_SubJob) {
+		/* Since subjob never go to F state,
+		 * check if any one setting it to F state. */
+		if (JOB_STATE_FINISHED == newstate)
+			newstate = JOB_STATE_EXPIRED;
+		update_subjob_state(pjob, newstate);
+	}
+
```

Comment:
In which scenario this block of code is required?

Reply:
This block is required for a subjob with below call hierarchy.
![image](https://user-images.githubusercontent.com/30307577/37912513-50c07876-3130-11e8-9f60-1afebc335710.png)

Reply:
My interest was not to know the callers of this function. Here we are changing the state from JOB_STATE_FINISHED  to JOB_STATE_EXPIRED and wanted to know the reason for this. And also doesn't this happen before we call the function  svr_saveorpurge_finjobhist. This is all what I wanted to know.

Reply:
The Call hierarchy will give the picture of the usage scenario. As we can see from it this function is called from many other places apart from svr_saveorpurge_finjobhit() as this function is used to update the history of the finished job. Hence even if it is handled before svr_saveorpurge_finjobhist, we need to handle it here. And since the state of a finished subjob is always X, we add the if condition to check if any one is trying to store the subjob state as finished and handle it accordingly.